docs/secrets.md

```diff
@@ -0,0 +1,62 @@
+Notification endpoints will require credentials in order to communicate with the various external services.
+
+We do not want users to include those credentials in plain text into Flux scripts. We need to provide a mechanism for users load secrets from a safe storage mechanism and provide those secrets to the endpoints.
+
+**Backing Store:** : There are many backing store possibilities, which need not be discussed here.  Any store that provides secure access for key/value lookup may be used by providing an implementation of the KeyLookup  interface in the flux repository.  
+
+**Fine Grained Authorization**:  keys will generally not be accessible in the flux language.  A secret-enabled function will accept meta-data about the secret (a key) and use it to look up the stored value in memory. The lookup will error if the key is not found or if the user is not authorized to use that key.  
+
+# **Design**
+
+## Internal Functions
+A function that requires secret lookup must: 
+
+- Be built-in to the language
+- Register an ExternalDependencies struct  (note: it will be straightforward and advantageous to refactor the `InjectFromDependencies` and similar functions into this pattern as well).   
+- Not return secrets under any circumstances (e.g. in error messages) 
+
+The external interface for secret lookup is: 
+```Go
+type SecretService interface {
+// LoadSecret retrieves the secret value v found at key k 
+// ctx is the current query context
+    LoadSecret(ctx context.Context,k string) (string, error)
+}
+```
+
+the internal function must be "secret aware" which means that it knows how to search its own input parameters for references to secrets, so that it can make the proper key-for-value substitution at the appropriate time (as close to when the key is used as possible)
+
+for example, a http.post function may reference a token in the header object by having a sub-object with the form  `{secretKey: <string>}`
+http.post(header: {"Content Type": "application-json", "ApplicationSecret": {secretKey: "FLUX_HTTP_POST_KEY"}}, url: "http://some/url/", data: json_enc)
+
+## Flux Package
+We will provide a `secrets` package that will help the user create an object with the required properties for secret lookup.  
+A helper function is useful because we may wish to change the properties of a secret request over time, and we don't want users to have to update their code unless strictly needed.  
+
+The vault package contains: 
+
+- `secrets.get(key: "FLUX_SLACK_ID")`: returns an object with all properties needed to make a request, e.g.,  `{userID: "11918", orgID: "19119384", secretKey: "my-secret"}`
```

Comment:
This is a bit confusing. Below here, you say it will be `vault.get` but up here it appears to be `vault.secrets.get`. Can this be clarified?

It also seems to me like the secrets package itself may work more as something that is influxdb specific. May I propose that we have this be `influxdata/influxdb/secrets` similar to how we're doing monitoring/alerts and we eschew any direct references to vault as anything other than an implementation detail?

Reply:
Agreed, we should not reference vault in the Flux code.